### ISSUE TITLE
Feature Request: “Rate limit” Pushes to control user behavior.

### ISSUE THREAD

**Issue by mrmarcsmith** [parse-community/parse-server#6067]
I would like to make a PR to slowly release a push notification to users over a variable period of time. 
Imaging a hot dog stand app that want to push out to nearby users “come buy a hot dog!”. 

Current way: the push is sent out to everyone about the same time and there is a VERY long line at the hotdog stand and lost customers because the hot dog stand can’t handle the volume.

**Describe the solution you'd like**
the push is released to batches of users over an hour (or 2 hours) giving a steady stream of customers without overwhelming the stand.

On a technical level this could also be used to prevent traffic spikes to your app that can overwhelm servers and cause a poor user experience.

**Describe alternatives you've considered**

I’ve considered writing a cloud job that sends pushes over time but those jobs can’t be access or controlled from the dashboard and I think this feature could benefit the parse community.

**Additional context**

**Comment by davimacedo** [comment]
It makes sense for me. Would the first step be implementing the Scheduled Pushes? Once implemented, this feature here would be much easier. We could just schedule the batches.

**Comment by mrmarcsmith** [comment]
That's the idea! my thought was to use the "query" value to break them up by ranged objectId. Since parse sorts by Numbers then Uppercase then Lowercase the half way point would be the end of Capital U (Uzzzzzzzzz)
0123456789ABCDEFGHIJKLMNOPQRSTU >> middle<< VWXYZabcdefghijklmnopqrstuvwxyz

so to break this query in half:
Original
```
{\"email\":\"mr.marcsmith@me.com\",\"deviceType\":{\"$in\":[\"ios\"]}}
```

we would get two batches:
Batch 1:
```
{\"objectId\":{\"$lt\":\"Uzzzzzzzzz\"},\"email\":\"example@example.com\",\"deviceType\":{\"$in\":[\"ios\"]}}
```

Batch 2:
```
{\"objectId\":{\"$gte\":\"Uzzzzzzzzz\"},\"email\":\"example@example.com\",\"deviceType\":{\"$in\":[\"ios\"]}}
```

A. Can you think of any reason the objectId field would be queried against except when we use direct equals (which would Never need batching)?

B. It doesn't look like the installation class has a default "user" pointer field. If I added one, instead of randomly batching by installation which might send to one user the same thing thought the day. we could guarantee the grouping all a user's devices in a single batch. while it's true that batching by user would unevenly distribute the actual number of push notifications (imaging one user with 10 devices), the whole goal of this is to evenly distribute the USERS, not the devices. 

Would anyone object to me adding the field "user" to the default _Installation schema AND grabbing the user that saved the _Installation (if present) for placement into the "user" field of the Installation Object?

**Comment by mtrezza** [comment]
@mrmarcsmith Maybe I'm missing something here, but `U` would only be the halfway point if the object IDs were distributed evenly across the alphabet? Also, executing multiple queries could be an issue in a collection where data changes fast, depending on your filter criteria.

Would it make sense to get all users who should receive a push in a single query, create an object in `_PushStatus` for each scheduled push, assign each push a batch number and then send them from there?

**Comment by mrmarcsmith** [comment]
Hi @mtrezza,  I believe I failed to communicate some of the basic concepts about my idea so let me try again with a more detailed fictional user story. 
Brad operates the hot dog booth at his local high school football games. He knows that the games last 2 hours and would like to maximize his sales during that time but a single push to all users would overwhelm his stand. Many students have the school app downloaded. The school’s app developer writes a function that triggers the following push at 6pm when the game starts
```
Parse.Push.send({
data: { “alert”:”Support your HS, purchase a delicious Hot Dog!”},
where: {userLocation: [nearby]},
rateLimit: 60 * 60 * 2 // this tells parse “ disperse sending this push over 2 hours”
})
```

That’s ALL the user has to do.
How do we accomplish this under the hood with Parse?

We intercept the saving of that one _PushStatus objects and break it up into multiple scheduled _PushStatus objects (see #6066). 

What method should we use to break up the _PushStatus object? We can’t run the query and use containedIn with the users objectIds because the app could have tens of thousands of users and storing them in the _PushStatus object would be a terrible idea. So instead we break up the objectIds into ranges, so 
At 6:00pm we send the push to all users who are nearby AND where the users objectId starts with 0
At 6:01pm nearby & objectId starts with 1
At 6:02pm nearby & 2
... 
At 6:59pm nearby & F
At 7:00pm nearby & G
...
At 7:59pm nearby & y
At 8:00pm nearby & z

Note: If a “z” user doesn’t arrive to the game until 7:59pm they will still get a push because the query is executed at the time the “z”  batch goes out (8pm)  and NOT at the time of original scheduling.
Likewise the opposite is true, if a “y” user leaves at 7:58 they will NOT receive a push because they no longer qualify for the “nearby” query param at 7:59pm when the “y” batch goes out.
